### DISCUSSION TITLE
Feature request thread!

### DISCUSSION THREAD

**Discussion by adc-ax** [adc-ax/GorkhonScriptEditor#3] [Ideas]
Kindly put your feature requests here

**Comment by RoSoDude** [comment]
1. Automatically remove "Function name" and "Number of arguments" when clicking on the fields above the Add Function button. If no text is entered and the player clicks away, fill the text back in
2. Add some space between the "Add function" UI elements and the "Append instruction" element
3. Add similar functionality to add strings (will have to distinguish between 8-bit and 16-bit, not sure how this works exactly)
4. When the user single clicks on a label or subroutine within the label/subroutine tab, this should open the expand its dropdown (or hide the dropdown if it's already expanded). If the user double clicks, then send them to that location (and don't close the dropdown!). Jumping to the address should probably expand it as well if it's not already open
5. When the user single/double clicks on the address for a label or subroutine from within the diassembled instruction tab, this should execute the same behavior in the label/subroutine tab
7. The current workflow for assembling scripts and exporting was not immediately intuitive to me. Exporting a script should first attempt to reassemble. Reassembling after adding a function appears to clear the added function; you have to export and then import to see the added function in the editor
8. Ideally, a warning should pop up if the script fails to assemble for any reason (currently GSE crashes)
9. If you click on "reassemble script" without first clicking out of the text box you were editing, any edits you made within that text box are lost. It should be assumed that you want to reassmble with the changes you made without first clicking out
10. There is a bug in which newly added functions don't always appear in the "Functions" tab. This can be fixed by tabbing back and forth?

**Comment by adc-ax** [comment]
> 1. Automatically remove "Function name" and "Number of arguments" when clicking on the fields above the Add Function button. If no text is entered and the player clicks away, fill the text back in

Will require extending WPF a bit, will figure out eventually, for now, a compromise will be put in place
<img width="148" alt="image" src="https://github.com/adc-ax/GorkhonScriptEditor/assets/47045905/06d39bbb-f1ac-4260-98cd-29834c1ae8c8">


> 2. Add some space between the "Add function" UI elements and the "Append instruction" element

Done, coming in the next version (may refine further before submission)

> 3. Add similar functionality to add strings (will have to distinguish between 8-bit and 16-bit, not sure how this works exactly)

The background piping is in place partially as of now, need to expose it on the UI + make sure everything works as intended.

> 4. When the user single clicks on a label or subroutine within the label/subroutine tab, this should open the expand its dropdown (or hide the dropdown if it's already expanded). If the user double clicks, then send them to that location (and don't close the dropdown!). Jumping to the address should probably expand it as well if it's not already open

Currently not focusing on this as much, but will eventually get to this.

> 5. When the user single/double clicks on the address for a label or subroutine from within the diassembled instruction tab, this should execute the same behavior in the label/subroutine tab

Currently not focusing on this as much, but will eventually get to this.

> 6. The current workflow for assembling scripts and exporting was not immediately intuitive to me. Exporting a script should first attempt to reassemble. Reassembling after adding a function appears to clear the added function; you have to export and then import to see the added function in the editor

Fixed

> 7. Ideally, a warning should pop up if the script fails to assemble for any reason (currently GSE crashes)

This is on the roadmap!

> 8. If you click on "reassemble script" without first clicking out of the text box you were editing, any edits you made within that text box are lost. It should be assumed that you want to reassmble with the changes you made without first clicking out

Looking into this now.

> 9. There is a bug in which newly added functions don't always appear in the "Functions" tab. This can be fixed by tabbing back and forth?

Yep, took me a while to figure out the reason behind this, apparently I didn't use an ObservableCollection where appropriate. Seems to be fixed in my local build as of today.
